### PR TITLE
Add query building API for the `$fill` aggregation pipeline stage

### PR DESCRIPTION
**Pull request by stIncMale**
The new API:

- `WindowedComputations.locf` for [`$locf`](https://www.mongodb.com/docs/v6.0/reference/operator/aggregation/locf/)
- `WindowedComputations.linearFill` for [`$linearFill`](https://www.mongodb.com/docs/v6.0/reference/operator/aggregation/linearFill/)
- `Aggregates.fill` for [`$fill`](https://www.mongodb.com/docs/v6.0/reference/operator/aggregation/fill/)

The server-side syntax document is [here](https://docs.google.com/document/d/1pwIBc5XzbZkMnZH-FAoVYf2OOUzQM2VT8wbtoC4FMTg/edit#heading=h.pkrftzkil1l1).

You may see how the API usage looks like, or play with it in

- `AggregatesFunctionalSpecification.'$setWindowFields'`
- `AggregatesFunctionalSpecification.'$fill'`

JAVA-4394

### CODE REVIEW

**Comment by stIncMale** [comment]
@katcharov, I added you in case you'd like to express your thoughts about the "debatable" part explained in the description of this PR. Feel free to ignore)

**Comment by stIncMale** [comment]
Moved back to "Ready for review".

**Comment by stIncMale** [comment]
The PR was refactored. A single pipeline stage specification us no longer used. Instead
- the API uses an object that represents optional fields (similarly to `Aggregates.bucket`, `Aggregates.search`; the API of this object has the same style as the API of `SearchOptions`);
- the API uses varargs to represents `output`, similarly to `Aggregates.setWindowFields`.

**Review by rozza — COMMENTED** [review]
This is a tricky API as the syntax seems to suggest optional values first - which is different to our (Java driver) general approach.

I found the name FillComputation surprising as it represents the output field in the syntax. Did you try any other names? I can only suggest FillOutput then type name is more closely linked to the field it represents.

**Comment by stIncMale** [comment]
> This is a tricky API as the syntax seems to suggest optional values first - which is different to our (Java driver) general approach.

This is only true if we want to accept `output` as varargs. This is not a new situation for us, and we had exactly the same situation in `Aggregates.setWindowFields` which, in turn, was modelled after `Aggregates.group`, that accepts its output via `varargs` (this part is not called `output` in the MQL, which only tells us that the MQL is also inconsistent with itself). One may also observe that `Aggregates.bucket` accepts its output via `BucketOptions.output`, which is different from `Aggregates.group`.

> I found the name FillComputation surprising as it represents the output field in the syntax. Did you try any other names? I can only suggest FillOutput then type name is more closely linked to the field it represents.

We had this exact discussion for `Aggregates.setWindowFields`, because `WindowedComputation` also seemed weird, yet `WindowedOutput`/`WindowOutput` appears to be even weirder. I constructed the name `FillComputation` to be similar to `WindowedComputation`. If we rename `FillComputation` to `FillOutput`, then we should also rename `WindowedComputation` to `WindowedOutput` (or `WindowOutput` 🤷‍♂️). The upcoming release is the last point where we can do this, because `Aggregates.setWindowFields` goes out of beta in 4.7.0.

Ross, is your proposal to rename `WindowedComputation`/`FillComputation` to `WindowedOutput`/`FillOutput`?

**Comment by rozza** [comment]
I agree the naming is difficult / problematic - sorry to rehash old ground. I noticed `setWindowFields` uses the param name `output` for `WindowedComputation`.

Looking at the naming:

* `WindowedComputation` - Relates to the context but uses new language.
* `WindowFieldOutput` - Relates to MQL but awkward to read.
* `WindowOutputField` - Relates to MQL
* `WindowOutput` - Relates to MQL but is very vague (meaningless?).

Personally I think I prefer: `FillOutputField` to `FillComputation` and `WindowOutputField`to `WindowedComputation`.

Regardless of naming I'll rereview in the morning. I'm also happy to delay naming decisions (as a separate ticket) until Jeff is back to see if we can get consensus.

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy [hunk 1]
```diff
@@ -1275,4 +1288,59 @@ class AggregatesFunctionalSpecification extends OperationFunctionalSpecification
...
+                new Document('partition', new Document('id', 10))
+                        .append('_id', 2)
+                        .append('doc', null),
+                new Document('partition', new Document('id', 20))
+                        .append('_id', 3)
+                        .append('field1', 3)
+                        .append('doc', new Document('field2', 3))]
+        getCollectionHelper().insertDocuments(docs)
+        String resultField = 'result'
+        LinkedList<Bson> stages = [
```
Comment (rozza) [inline driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy:1310]: Q - why linkedlist over just List?

**Review by rozza — APPROVED** [review]
LGTM!

I had one question that is just for my learning - no action needed.

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy (hunk 1 above)
Comment (stIncMale) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy]: Indeed, I could have get by with the `List` interface. Done.

**Review by katcharov — APPROVED** [review]
lgtm

Just a trivial doc suggestion.

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 2]
```diff
@@ -656,6 +661,60 @@ public static <TExpression> Bson setWindowFields(@Nullable final TExpression par
...
+     * @mongodb.driver.manual reference/operator/aggregation/fill/ $fill
+     * @mongodb.server.release 5.3
+     * @since 4.7
+     */
+    public static Bson fill(final FillOptions options, final FillComputation output, final FillComputation... moreOutput) {
+        return fill(options, concat(notNull("output", output), moreOutput));
+    }
+
+    /**
+     * Creates a {@code $fill} pipeline stage, which sets values to fields when they are {@link BsonType#NULL Null} or missing.
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/Aggregates.java]: Done. This change was also needed in the Scala code.

**Pull request merged**